### ISSUE TITLE
Inference with nnunetv2 models trained with specific trainers

### ISSUE THREAD

**Issue by plbenveniste** [spinalcordtoolbox/spinalcordtoolbox#4955]
# Description

I am opening this issue to centralise the discussion regarding how to perform inference with nnunetv2 models which were trained with specific trainers. The problem comes from these few lines of code: 
```python
 # restore network
num_input_channels = determine_num_input_channels(plans_manager, configuration_manager, dataset_json)
trainer_class = recursive_find_python_class(join(nnunetv2.__path__[0], "training", "nnUNetTrainer"),
                                            trainer_name, 'nnunetv2.training.nnUNetTrainer')
if trainer_class is None:
    raise RuntimeError(f'Unable to locate trainer class {trainer_name} in nnunetv2.training.nnUNetTrainer. '
                       f'Please place it there (in any .py file)!')
network = trainer_class.build_network_architecture(
    configuration_manager.network_arch_class_name,
    configuration_manager.network_arch_init_kwargs,
    configuration_manager.network_arch_init_kwargs_req_import,
    num_input_channels,
    plans_manager.get_label_manager(dataset_json).num_segmentation_heads,
    enable_deep_supervision=False
)
```
If the trainer class isn't inside the installed nnunetv2 library, then the inference will crash.
As discussed previously, adding the trainer files for each model stored in SCT could be a (dirty) workaround: but this solution is not very clean and could cause some issues. 

# Questions raised:
- Why not re-code the entire inference pipeline? 
    Loading the model and performing inference aren't the tricky part. The major problems comes from al the pre-processing steps which are model specific and stored in the `plans.json` file.  Here is an example: [link](https://github.com/Luugaaa/nnunet_extracted_inference/blob/3d39eb675a22b413388b4dc4ab524047ba4cd555/inference_model/metadata/plans.json#L48)

# Ideas from @Luugaaa:

## Idea 1: 
We could change the code so that at inference we only use the nnUNetTrainer (the original one). This should have a limited impact as the trainer class at inference is only used to build the correct architecture through the following pieces of code: 
```python
network = trainer_class.build_network_architecture(
```
Which does the following: 
```python
def build_network_architecture(architecture_class_name: str,
                                   arch_init_kwargs: dict,
                                   arch_init_kwargs_req_import: Union[List[str], Tuple[str, ...]],
                                   num_input_channels: int,
                                   num_output_channels: int,
                                   enable_deep_supervision: bool = True) -> nn.Module:
        """
        This is where you build the architecture according to the plans. There is no obligation to use
        get_network_from_plans, this is just a utility we use for the nnU-Net default architectures. You can do what
        you want. Even ignore the plans and just return something static (as long as it can process the requested
        patch size)
        but don't bug us with your bugs arising from fiddling with this :-P
        This is the function that is called in inference as well! This is needed so that all network architecture
        variants can be loaded at inference time (inference will use the same nnUNetTrainer that was used for
        training, so if you change the network architecture during training by deriving a new trainer class then
        inference will know about it).

        If you need to know how many segmentation outputs your custom architecture needs to have, use the following snippet:
        > label_manager = plans_manager.get_label_manager(dataset_json)
        > label_manager.num_segmentation_heads
        (why so complicated? -> We can have either classical training (classes) or regions. If we have regions,
        the number of outputs is != the number of classes. Also there is the ignore label for which no output
        should be generated. label_manager takes care of all that for you.)

        """
        return get_network_from_plans(
            architecture_class_name,
            arch_init_kwargs,
            arch_init_kwargs_req_import,
            num_input_channels,
            num_output_channels,
            allow_init=True,
            deep_supervision=enable_deep_supervision)
```

## Idea 2:

To prevent issues with the loading of the model building of the right architecture, we could instead export the nnunetv2 model with a torch script. This way the model is loadable with torch.jit.load and its underlying architecture doesn't need to be found within the nnunet code nor a trainer file.

@Luugaaa Don't hesitate to correct me if I am wrong.

**Comment by Luugaaa** [comment]
@plbenveniste Thanks for creating the issue! From what I understand, this all looks correct. 🙏

Hey @mathieuboudreau and @joshuacwnewton, I was told you could help me understand the SCT codebase better, since I'm not familiar with it at all.

### Current Status

First off, I've tested **Solution 1** and it seems to work well.

I haven't tested **Solution 2** yet. We're a bit concerned about potential incompatibilities between some nnUnet processing and a TorchScript model, but I'll check on that asap.

---

### Which solution should we go with?

Assuming both solutions work, I'd love to get your opinion on which one we should implement.

My understanding is that nnUnet is used as a package and isn't deeply integrated into the codebase, and the plan is to keep it that way to make updating nnUnet easier (is that right?)

Here's a breakdown of the two solutions I propose :

* **Solution 1:**
    * We could add a new inference directory.
    * The code in this directory would depend on the `nnunetv2` package but not on the `CustomTrainer` class (which is the whole point).
    * It would essentially be a modified copy of a part of the `nnunetv2` package, avoiding any direct changes to the package itself.

* **Solution 2:**
    * This would give us more independence from the nnUnet framework by saving the architecture inside the checkpoint.
    * This would require either:
        * Changing the nnUnet package code directly (both in the inference and `save_checkpoint` code).
        * **OR**
        * Creating a new customized inference code (like in Solution 1) and overriding the [`nnUnetTrainer.save_checkpoint()` function](https://github.com/MIC-DKFZ/nnUNet/blob/f1851fbaf2c53dcb51b079b60a01de528a7d0c17/nnunetv2/training/nnUNetTrainer/nnUNetTrainer.py#L1149) within a `CustomTrainer` that would always need to be used as base for any training.
    * Another sub-solution, proposed by PL, is to keep the `save_checkpoint` code as-is and use a post-training script to load the model with nnU-Net and then save it with TorchScript. This would basically add an "export" script as an extra step after training.

Let me know what you think when you have a moment. Any feedback on what we should implement, how we should do it, or any other thoughts are welcome!

**Comment by joshuacwnewton** [comment]
Thanks for opening this issue, and for highlighting the relevant lines in the `nnunetv2` package! Just to make sure we're not missing something (I doubt we are), I've also raised the root issue with the `nnunetv2` devs so that they are aware that their hardcoding of the `nnunetv2` directory is causing issues:

- https://github.com/MIC-DKFZ/nnUNet/issues/2856

**Comment by joshuacwnewton** [comment]
My apologies, but separately from @Luugaaa's ideas, didn't we discuss in the [2025-07-03 SCT meeting](https://docs.google.com/document/d/10txnIVrX3roF-VYP3yYUVLk7KmeL-43-ppKMKJ1BCpE/edit?tab=t.lr9hw84c0h15#heading=h.rdxbbazcgupl) the possibility of having a `.py` file with the trainer code, then copying the `.py` file to the `nnunetv2` directory (as described):

```python
if trainer_class is None:
    raise RuntimeError(f'Unable to locate trainer class {trainer_name} in nnunetv2.training.nnUNetTrainer. '
                       f'Please place it there (in any .py file)!')
```

Relevant meeting minutes:

<img width="587" height="150" alt="Image" src="https://github.com/user-attachments/assets/bd699cc8-4d4e-4724-9e67-df951a968d01" />

Did we rule this idea out? Maybe I'm misunderstanding something...

**Comment by jcohenadad** [comment]
> My apologies, but separately from @Luugaaa's ideas, didn't we discuss in the [2025-07-03 SCT meeting](https://docs.google.com/document/d/10txnIVrX3roF-VYP3yYUVLk7KmeL-43-ppKMKJ1BCpE/edit?tab=t.lr9hw84c0h15#heading=h.rdxbbazcgupl) the possibility of having a .py file with the trainer class definition, then copying the .py file to the nnunetv2 directory when installing the model or running inference (as instructed by the nnunetv2 devs):

Absolutely-- excellent point @joshuacwnewton. I asked @plbenveniste to open an issue to reflect what was discussed during our SCT meeting (if bits are missing, they should be added indeed), so that this (important) discussion does not only appear in our internal SCT meeting docs, because (i) it also concerns non-SCT projects (eg ADS) and (ii) ppl not present during our internal meeting (eg @Luugaaa at that time) could not contribute to it. So again, the idea of this issue is for posterity and for continuing an active discussion within the Github ecosystem, which is more visible, allows cross-referencing, etc.

**Comment by mguaypaq** [comment]
Amazing investigation @joshuacwnewton, thanks!

**Comment by Luugaaa** [comment]
@plbenveniste I'm very sorry for this late reply. I overrode the nnUnetTrainer load_checkpoint function with [this one](https://github.com/Luugaaa/nnUNet/blob/c35a88320de78ed329b61cc3591c94aaefc798ec/nnunetv2/training/nnUNetTrainer/variants/nnUNetTrainerWandb.py#L108) though a custom trainer. That led to the error I mentioned above.

**Comment by Luugaaa** [comment]
During a brief exchange with @mguaypaq at the lab, we discussed the idea of making an inference code based on nnunetv2 but that would be decoupled from the package itself. This code would then be shipped with the checkpoint and metadata. The goal here is not only to have an inference code that is trainer agnostic but also to improve compatibility and reproducibility by removing the nnunetv2 package from the equation. Of course, if nnUnet changes drastically, this inference code would need to be updated for the new models, but there is a high chance this inference code will always work, as long as the processing classes are kept up to date (see the [readme](https://github.com/Luugaaa/nnUNet/tree/dev/paul/extracted_inference?tab=readme-ov-file#maintenance--extensibility)). Other than the classes, in the future if nnUnet changes the format of the metadata produced along with the model, that could be an issue.  

I am mostly done with this [implementation](https://github.com/Luugaaa/nnUNet/tree/dev/paul/extracted_inference), please give it a look when you have time if you think that can be useful. You'll find some details in the readme. I made it based on the nnunet repo, so that you can more easily examine the changes, however I did move some files, so some changes might not be clearly showed. I tested it on 3 models trained with different custom trainers, it ran successfully. 

Ps : I removed the recursive function that look for classes in the codebase and replaced that with something that is more maintainable in my opinion)

**Comment by joshuacwnewton** [comment]
> I am mostly done with this [implementation](https://github.com/Luugaaa/nnUNet/tree/dev/paul/extracted_inference), please give it a look when you have time if you think that can be useful.

At first I felt a little bit wary about this implementation?

<details>
<summary> Thoughts... </summary>

The first thing that stands out to me is the number of extracted submodules from `nnunetv2`. There is still quite a lot of code that we are extracting, so much so that this feels to me like a full fork of `nnunetv2`.

Even if this code works, unless we keep our fork up to date, we will miss out on any bugfixes or issues that the upstream `nnunetv2` devs identify. I looked at the git history for the modules we extracted, and there still seems to be a fair amount of activity over the past few months?

</details>

But then I realized that SCT already pins a specific version of `nnunetv2`, so my same worries apply to what we're already doing. (In other words, any bugs/issues in the extracted code will _also_ exist in old pinned versions of `nnunetv2`, too.)

So, I think I support this effort! :tada:

----

> This code would then be shipped with the checkpoint and metadata. 

Oh, interesting! So, there would be no `nnunetv2` code inside of SCT, and instead we would treat `nnunetv2` in a similar way as we do our C++ binaries -- a black box that we call via a subprocess that takes in image filenames and outputs image filenames.

I like this a lot! By coupling the inference code with the model files themselves, we ensure that each model `.zip` has everything that's needed to run inference, without having to worry about version conflicts between different models.

I do have a few questions, though:

- Right now, we're performing inference using the [`segment_nnunet`](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/ae939b1c3be5f9def829562c82a45cb9dda8dabb/spinalcordtoolbox/deepseg/inference.py#L222-L341) function. Since this function accepts `path_img` and returns `fnames_out`... I guess we would create a drop-in replacement for this function where we pass `path_img` to whatever inference script is bundled with the model, then return the same outputs?
- How will we handle _`nnunetv2`'s_ dependencies? I remember in the past that we sometimes had issues with `nnunetv2`'s other packages that were maintained by the same developers (e.g. `acvl_utils`: https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4679) as well as interactions with other packages in the torch ecosystem. So, I'm wondering how will we define the required dependencies so that we can make sure they're installed for the extracted code to run?
   - Maybe we could create a `requirements.txt` file inside our fork and distribute that alongside the inference code? Then, we try to keep SCT's `requirements.txt` file in sync so that SCT installs the same dependencies?
   - Maybe we create a simple test that downloads each model, grabs its requirements files, then checks them against SCT's requirements somehow? Just to make sure that any PRs to SCT's requirements don't end up incompatible with the requirements of the inference code.

**Comment by Luugaaa** [comment]
@joshuacwnewton thank you, happy to help ! 
yes I completely agree with you, this inference implementation is really like a fork of unet. Unfortunately, I don't think we can create an inference code that would be preprocessor/normalization/image reader agnostic, so we still need to keep at least those in our inference code 😓

To answer your question to the best of my knowledge:
* I would say that it can be easily done. Currently, the inference script takes in a dir_path in which it will iterate through the available files and run inference. So it should only be a slight modification. However, my current implementation relies on nnUnet code, which requires a specific file naming format. Is that the same with the  [segment_nnunet](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/ae939b1c3be5f9def829562c82a45cb9dda8dabb/spinalcordtoolbox/deepseg/inference.py#L222-L341) function ?
* ah yes, I see. I think the dependencies versions are very likely to change with nnunet versions, even slightly. That will not necessarily create incompatibilities, but a code checking the inference requirements against the SCT requirements will probably often find some differences. Maybe having a requirements file shipped with the model and the code is the best solution. I believe we can install those specific packages in a local_packages dir inside the inference dir, so that it doesn't interfere with SCT's packages. I'm not very experienced with those software engineering kinds of things, so the others could probably give a better input than me.
